cmd/alertmanager/main.go

```diff
@@ -304,30 +279,24 @@ func run() int {
 		return d + waitFunc()
 	}
 
-	var hash float64
-	reload := func() (err error) {
-		level.Info(logger).Log("msg", "Loading configuration file", "file", *configFile)
-		defer func() {
-			if err != nil {
-				level.Error(logger).Log("msg", "Loading configuration file failed", "file", *configFile, "err", err)
-				configSuccess.Set(0)
-			} else {
-				configSuccess.Set(1)
-				configSuccessTime.Set(float64(time.Now().Unix()))
-				configHash.Set(hash)
-			}
-		}()
-
-		conf, plainCfg, err := config.LoadFile(*configFile)
-		if err != nil {
-			return err
-		}
+	var (
+		inhibitor *inhibit.Inhibitor
+		tmpl      *template.Template
+		pipeline  notify.Stage
+		disp      *dispatch.Dispatcher
+	)
 
-		hash = md5HashAsMetricValue(plainCfg)
+	defer disp.Stop()
 
+	configCoordinator := config.NewCoordinator(
+		*configFile,
+		prometheus.DefaultRegisterer,
+		log.With(logger, "component", "configuration"),
+	)
+	configCoordinator.Subscribe(func(conf *config.Config) {
 		tmpl, err = template.FromGlobs(conf.Templates...)
 		if err != nil {
-			return err
+			level.Error(logger).Log("err", fmt.Errorf("failed to parse templates: %v", err.Error()))
```

Comment:
If we enter this branch, the error wouldn't bubble up anymore. Is that intended? Wouldn't it make sense for `Coordinator.Subscribe` to accept `func(*Config) error`?

Reply:
Right, that got me thinking as well. If `Coordinator` accepts `func(*Config) error`, it would need to handle errors returned by other components logic. What should it do with such error? It can't know what kind of errors are possible, hence it doesn't know how to handle them, right? How would you handle the errors in the `Coordinator` code?

Reply:
In the Prometheus code, we apply all "reloaders" (equivalent to subscribers), log an error message  for any that fails and return a generic error message tot the caller in case one reloader failed.

https://github.com/prometheus/prometheus/blob/e4a741cb7dfa24978097971c2c8d5c8cc12e6cef/cmd/prometheus/main.go#L714-L744

Reply:
Ok, good point. Did the relevant changes. Instead of returning a generic error message, I am returning and logging a specific one. Let me know if that is ok for you. Should be better for debugging.